cs1/tokenbucket.py

```diff
@@ -0,0 +1,29 @@
+import time
+
+
+class TokenBucket:
+
+    def __init__(self, tokens, time_unit):
+        self.tokens = tokens
+        self.time_unit = time_unit
+        self.bucket = tokens
+        self.last_check = time.time()
+
+    def consume(self):
+        if (self.bucket < 1):
+            time.wait(self.time_unit / self.tokens)
+            self.bucket = self.bucket + 1
+
+        if (self.bucket > self.tokens):
+            self.bucket = self.tokens
+        else:
+            self.bucket = self.bucket - 1
+
+    def put_token(self):
```

Comment:
Why make these separate functions? From an API perspective, a user of this class is concerned with one thing: determining whether they've hit a limit. If this is part of the API surface, that requires the user to do something orthogonal to that purpose (in this case, giving themselves more quota).

I think the public API should just be one function (e.g., your `consume`) and that function can add back quota before taking from it.

Reply:
I would even suggest making this class return a boolean -- whether the limit is exceeded or not, and have the caller decide how to back off.